File: chapter-11/ios-specific/NativeModuleApp/ios/NativeModuleApp/HelloManager.m
//
//  HelloManager.m
//  NativeModuleApp
//
//  This Objective-C file registers the Swift native module with React Native's bridge.
//  Since React Native's module system is built in Obj-C, Swift modules need this
//  companion file to declare their existence to the bridge.
//
//  RCT_EXTERN_MODULE declares the module name (must match @objc(HelloManager) in Swift)
//  RCT_EXTERN_METHOD declares each method signature exposed to JavaScript
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(HelloManager, NSObject)

RCT_EXTERN_METHOD(
  greetUser:(NSString *)name
  isAdmin:(BOOL)isAdmin
  callback:(RCTResponseSenderBlock)callback
)

@end

File: chapter-11/ios-specific/NativeModuleApp/ios/NativeModuleApp/HelloManager.swift
import Foundation

/**
 * HelloManager is a custom native module that demonstrates how to expose
 * Swift methods to JavaScript via the React Native bridge.
 *
 * In JavaScript, this module is accessed via:
 *   import { NativeModules } from 'react-native';
 *   const { HelloManager } = NativeModules;
 *   HelloManager.greetUser(name, isAdmin, callback);
 *
 * Key points:
 * - The class must extend NSObject (required by the Obj-C runtime bridge)
 * - @objc attribute exposes methods to Objective-C (which RN's bridge uses)
 * - The companion .m file (HelloManager.m) uses RCT_EXTERN_MODULE and
 *   RCT_EXTERN_METHOD to register the module with React Native's bridge
 */
@objc(HelloManager)
class HelloManager: NSObject {

  /**
   * Any method exposed to JS must be annotated with @objc.
   * Parameters are automatically converted between JS and Swift types.
   * RCTResponseSenderBlock is the callback type that sends data back to JS.
   */
  @objc
  func greetUser(
    _ name: String,
    isAdmin: Bool,
    callback: @escaping RCTResponseSenderBlock
  ) {
    let adminStatus = isAdmin ? "are" : "are not"
    let greeting = "Welcome \(name), you \(adminStatus) an administrator."
    callback([greeting])
  }

  /**
   * Tells React Native that this module's methods should be called on the
   * main thread. Return false if your methods don't touch UI.
   */
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
